server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java

```diff
@@ -338,6 +345,9 @@ public void writeTo(StreamOutput out) throws IOException {
                 out.writeCollection(knnSearch);
             }
         }
+        if (out.getTransportVersion().onOrAfter(TransportVersion.V_8_8_0)) {
+            out.writeOptionalNamedWriteable(rankBuilder);
+        }
```

Comment:
I would do this for safety in all places where we are serializing additional info. in some internal places it could even be an assert though probably.

Reply:
I do not think we should have validation exceptions like that in the middle of serializing objects. This is because I think we should move towards less special casing inside StreamInput/Output code, so that we could eventually generate the code base on simple POJOs (or better yet, records). The writeTo and StreamInput constructors are meant as simple serialization methods, not validators.

There are two scenarios you described: mixed cluster and older ccs.

For mixed cluster, I do not think we should do any special checking. That scenario means a user is sending a request asking for rrf *in the middle of upgrading to 8.8*. It's an edge case which we protect users from by suggesting first upgrading clusters before clients.

For CCS, the scenario is valid. However, we should be able to check such scenarios elsewhere, sending back an appropriate 400, not an internal error because we failed to write a stream. I'm not sure the best place to do that, but perhaps it would be in `TransportSearchAction.executeRequest`, similar to the `checkCCSVersionCompatibility` call there?

Reply:
We have switched to a different pattern of throwing exception a while ago, and I thought it was a discussion we had with the overall ES team. This came out of the CCS working group I believe, and it affects both mixed clusters and CCS, where CCS is like you identified the most prominent case.

In general, we don't want to accept requests that are asking for things that we can't accommodate. We did it for runtime fields: if you provide for runtime mappings in a search request, we will fail serializing the request to an older version node that does not support it, instead of doing what the user asked only on nodes that support it and silently failing on older nodes. The checkCCSVersionCompatibility builds on it as it relies on the serialization layer throwing exception whenever something that the previous version does not support gets serialized.

@DaveCTurner @nik9000 I remember you two were involved in these discussions too, and I initially had a similar reaction as Ryan, but then I got convinced and I adopted the "new approach". Do you have more context to it?

Reply:
I definitely prefer to detect the use of "future" features, and reject them, before we even get to the point of trying to serialize a message that the remote node won't understand. We can give much more useful feedback (and avoid unnecessary work) if we do.

But we still have to do _something_ on the serialization path when we encounter something that can't be translated into an older protocol version. Either we bail out with an exception, or we just proceed blindly and send an invalid message. Given this choice, I think we should prefer to bail out wherever possible. Receiving an invalid message trips assertions and throws an exception (on the sender) but there are cases where the message remains syntactically valid, yet differs semantically from what we intended, and that's a bug.

If we have upstream checks that prevent it then we can certainly assert that this branch never happens too, but still IMO we should throw an exception after the `assert false` just in case we're missing test coverage.

Reply:
I understand the concern, and agree it is better to bail. Adding assertions/throw assertions in the else case sounds good. My main ask is to keep validation separate from serialization. If we can move such a check earlier in the request, such as in `TransportSearchAction.executeRequest`, I think it will make the code easier to reason about too.

Reply:
@javanna I went ahead and added `IllegalArgumentException`s at the different places where we check for 8.8. Would you please point me to where runtime fields does this check? I see the one for CCS, but I don't see one for just a standard mixed cluster.

Reply:
This is the check that I meant: https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java#L321